Shuffle depends instructions and recommend modern make for macOS

Sjors opened the pull request:
macOS ships with GNU Make 3.81 from 2006. This has caused
difficult to debug issues, e.g. #32070 and #30978.

Tell users / developers who use the depends system to install a modern version of `make`.

This PR does not change the non-depends build.

Although Homebrew allows overriding the system `make`, we instead just instruct users to build with `gmake`. This way there should be no impact on other projects they wish to compile.

To increase the likeliness of anyone actually seeing and following this instruction, the first commit moves things around in `depends/README.md`. It now starts with instructions for a local build and moves cross-compilation to the end. For each platform it shows what to install (`apt install`, `brew install`, etc) and what command to run (`make` or `gmake`).

There previously was no macOS specific section, so this is added. It points to the general `build-osx.md` for how to install the Xcode Command Line Tools and Homebrew Package Manager.

I didn't test on an empty system.

Preview: https://github.com/Sjors/bitcoin/tree/2025/03/mc-make/depends#depends-build

Sjors committed:
doc: shuffle depends instructions

Native compilation is explained before cross-compilation. Move
install and (g)make steps up.

In the Configuring section, use Linux native compilation as the
example instead of Windows cross-compile.

hebasto commented:
> Tell users / developers who use the depends system to install a modern version of `make`.

Concept ACK on that.

willcl-ark reviewed (approved):
ACK 3ad36fce4be0aa413639468118d6dbd7379a2106

I confirm that the `make` brew formula [symlinks itself](https://github.com/Homebrew/homebrew-core/blob/5c4e0f50fa0cee5ae512f09ae887e3f5d0c05318/Formula/m/make.rb#L52-L55) to `gmake`.

Did a successful (`NO_QT=1`) depends build using `gmake` on macOS 15 and no longer see the errors from #30978.

I think it would be good to keep #30978 open if this is merged, to address https://github.com/bitcoin/bitcoin/issues/30978#issuecomment-2582535849

Sjors commented on depends/README.md:
```diff
@@ -1,16 +1,59 @@
+### macOS
 
-    make HOST=host-platform-triplet
+Install Xcode Command Line Tools and Homebrew Package Manager,
+see [build-osx.md](../doc/build-osx.md).
```
The Required Dependencies would defeat the purpose of depends. at least `boost` and `libevent`. But maybe `pkgconf` and `cmake` should be added to the brew install incantation?

willcl-ark replied:
Right, I meant those two only and would agree with adding them to the incantation here, if you agreed.

Sjors replied:
Added

Sjors replied:
Apparently `pkgconf` isn't needed, see https://github.com/bitcoin/bitcoin/pull/32086#discussion_r2000262395

willcl-ark reviewed (approved):
reACK 31a0c5f3caa80c8e6baf4e8f909a673442a852d5

fanquake commented on depends/README.md:
```diff
@@ -1,16 +1,59 @@
-### Usage
+# Depends build
+
+This is a system of building and caching dependencies necessary for building Bitcoin,
```
s/Bitcoin/Bitcoin Core/ 

> either on your local machine or through cross-compilation

I think this needs to be rephrased, or just dropped.

Sjors replied:
I pushed a variation.

fanquake commented on depends/README.md:
```diff
@@ -1,16 +1,59 @@
+see [build-osx.md](../doc/build-osx.md).
 
-For example:
+    brew install cmake pkgconf make
```
Do we need to copy-paste things here, given you've already linked to somewhere else that already has the same info? In either case, `pkgconf` isn't needed with depends, so should be dropped.

Sjors replied:
The main install instructions pull in libevent, boost (and pkgconf) which are not needed for a depends build. That's not going to be obvious, as demonstrated by the fact that I didn't know `pkgconf` isn't needed :-) (which I'll drop)

willcl-ark replied:
Unrelated to the changes in this PR, but as I was on a fresh `bookworm` docker image (sha256:18023f131f52fc3ea21973cabffe0b216c60b417fd2478e94d9d59981ebba6af), it appears that depends build fails to build bdb on Debian when following current instructions:

```bash
# docker run -it debian:bookworm /bin/bash

apt update
apt install -y git
git clone --depth=1 https://github.com/bitcoin/bitcoin
cd bitcoin/depends
apt install -y cmake curl make patch
make -j10 NO_QT=1

# build logs
Fetching db-4.8.30.NC.tar.gz from https://download.oracle.com/berkeley-db
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 21.7M  100 21.7M    0     0  15.5M      0  0:00:01  0:00:01 --:--:-- 15.5M
/bitcoin/depends/work/download/bdb-4.8.30/db-4.8.30.NC.tar.gz.temp: OK
Extracting bdb...
/bitcoin/depends/sources/db-4.8.30.NC.tar.gz: OK
Preprocessing bdb...
patching file dbinc/atomic.h
patching file mp/mp_fget.c
patching file mp/mp_mvcc.c
patching file mp/mp_region.c
patching file mutex/mut_method.c
patching file mutex/mut_tas.c
Configuring bdb...

<snip>

checking how to run the C++ preprocessor... /lib/cpp
configure: error: in `/bitcoin/depends/work/build/x86_64-pc-linux-gnu/bdb/4.8.30-a3c72735bb6/build_unix':
configure: error: C++ preprocessor "/lib/cpp" fails sanity check
See `config.log' for more details.
make: *** [funcs.mk:303: /bitcoin/depends/x86_64-pc-linux-gnu/.bdb_stamp_configured] Error 1

<snip>

# This is fixed by installing g++
apt install -y g++
```

Sjors replied:
> Unrelated to the changes in this PR

Indeed I'd like to avoid changing existing instructions here.

Sjors commented:
@laanwj in the longer run I hope we can "just" move depends over to cmake entirely.

hebasto commented:
> i'm hesitant about recommending people to use GNU tools on BSD variants, one of the advantages i hoped cmake would have is that we can just use native build tools as-is.

That's true.

However, our depends build subsystem is not managed by CMake and requires GNU Make, not BSD Make, to run.

laanwj commented:
> @laanwj in the longer run I hope we can "just" move depends over to cmake entirely.
> However, our depends build subsystem is not managed by CMake and requires GNU Make, not BSD Make, to run.

Ah yes *for the depends* it's unavoidable right now, makes sense, sorry.

Sjors commented:
@laanwj I clarified the PR description by adding "This PR does not change the non-depends build."

willcl-ark reviewed (approved):
reACK 3dd8223d44928a7e7c2cdc451655a8ef7d1ef5be

I've only tested the revised macOS instructions on a macOS system that already had other packages installed on it (not fresh), as I don't have a fresh machine to test on currently.

Sjors committed:
doc: recommend modern make for macOS depends

macOS ships with GNU Make 3.81 from 2006. This has caused
difficult to debug issues, see e.g.
9157d9e449870851ef455e077249ac46fc2df24c.

Also add ninja, needed since qt6.

Sjors commented:
Rebased after #30997, also mentions ninja for macOS native builds.

hebasto reviewed (commented):
In the following sentence:
https://github.com/bitcoin/bitcoin/blob/93d47a1143e47bbf66924a9b1eed16a8fa8c2ff5/depends/README.md?plain=1#L61

"the example" is no longer "above".

hebasto commented on depends/README.md:
```diff
@@ -1,16 +1,59 @@
+see [build-osx.md](../doc/build-osx.md).
 
-For example:
+    brew install cmake make ninja
```
Tested on macOS Sequoia 15.3.2 (Intel) -- looks good.

hebasto commented on depends/README.md:
```diff
@@ -1,24 +1,107 @@
+
+To build dependencies for the current arch+OS:
+
+    make
```
`gmake` is used in `doc/build-freebsd.md`.

hebasto reviewed (approved):
ACK ace66e125c01448a4bde0889b044544937bad6ea.

Sjors committed:
doc: recommend gmake for FreeBSD

Consistent with doc/build-freebsd.md

Sjors commented:
Rebased just in case and added a commit to switch FreeBSD recommendation to `gmake`.

hebasto reviewed (approved):
re-ACK 7238f242d78f3d71834764031d7904d19525bab3.

willcl-ark reviewed (approved):
reACK 7238f242d78f3d71834764031d7904d19525bab3

Changes since my last ACK (excl. the rebase) look OK to me.

fanquake commented on depends/README.md:
```diff
@@ -1,24 +1,107 @@
 order for it to pick up libraries, tools, and settings from the depends build,
 you must specify the toolchain file.
-In the above example, a file named `depends/x86_64-w64-mingw32/toolchain.cmake` will be
+In the above example for macOS, a file named `depends/aarch64-apple-darwin24.4.0/toolchain.cmake` will be
```
Please don't add triples with trailing versions numbers, as it's just something else to be "bumped". You can leave the windows example as-is.

Sjors replied:
There's no Windows example above, but I'll switch it to Linux - which doesn't use a version.

Sjors replied:
https://github.com/bitcoin/bitcoin/compare/7238f242d78f3d71834764031d7904d19525bab3..22cff32319de64cb98e1c89b9a7ed35611e89e27

fanquake replied:
Not sure what you mean. There's no need to change anything here, the Windows example was fine before your PR, and is still fine now.

Sjors replied:
The commit message of 99e6490dc51adde35b58e8d193aca7c1c422dbf3 explains why:

> In the Configuring section, use Linux native compilation as the
> example instead of Windows cross-compile.

Previously this documented started with cross compilation, which "above" referred to. In this PR I start with native compilation.

maflcko reviewed (commented):
review ACK 22cff32319de64cb98e1c89b9a7ed35611e89e27 🏣

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 22cff32319de64cb98e1c89b9a7ed35611e89e27 🏣
vKs0MrcJheMiMJ8wEAuqB0JgFjIna+8G3RL3GHyggqHGwWhDnPURKpp3ql3G+/Uf648kjhBQt47uUelVG4OGDg==
```

</details>

hebasto reviewed (approved):
re-ACK 22cff32319de64cb98e1c89b9a7ed35611e89e27.

willcl-ark reviewed (approved):
ACK 22cff32319de64cb98e1c89b9a7ed35611e89e27
